### PR TITLE
Extract Method Bug With Method Directly in Namespace

### PR DESCRIPTION
**Pull request by akhera99**
#44214

Hello!

I worked on a bug in which, if you paste a method directly into a namespace and highlight an expression in the method, ExtractMethod crashes. To fix this, I look at the ```GetValidSelectionAsync``` methods in C# and VB. This calls ```GetInitialSelectionInfo()``` which returns valid or invalid selections. I added a check to see if the highlighted expression is contained within some block that methods could be contained in, Classes, Records, Structs, Modules, etc. I was going to make this change in ```UnderValidContext()``` but the existing code works to short-circuit if the higlighted code is an ExpressionSyntax.

Please let me know if there is a better way to design this that fits with the design of the existing extractmethod code better!

### CODE REVIEW

File: src/Features/VisualBasic/Portable/ExtractMethod/VisualBasicSelectionValidator.vb [hunk 1]
```diff
@@ -494,6 +494,18 @@ result.ReadOutside().Any(Function(s) Equals(s, local)) Then
                        }
             End If
 
+            Dim commonRootContainer = commonRoot.FirstAncestorOrSelf(Of SyntaxNode)(Function(node) node.Kind = SyntaxKind.ClassBlock Or node.Kind = SyntaxKind.StructureBlock Or node.Kind = SyntaxKind.ModuleBlock)
+
+            If (commonRootContainer Is Nothing) Then
```
Comment (Youssef1313) [inline src/Features/VisualBasic/Portable/ExtractMethod/VisualBasicSelectionValidator.vb:499]:
```suggestion
            If commonRootContainer Is Nothing Then
```

File: src/Features/VisualBasic/Portable/ExtractMethod/VisualBasicSelectionValidator.vb [hunk 2]
```diff
@@ -494,6 +494,18 @@ result.ReadOutside().Any(Function(s) Equals(s, local)) Then
                        }
             End If
 
+            Dim commonRootContainer = commonRoot.FirstAncestorOrSelf(Of SyntaxNode)(Function(node) node.Kind = SyntaxKind.ClassBlock Or node.Kind = SyntaxKind.StructureBlock Or node.Kind = SyntaxKind.ModuleBlock)
```
Comment (Youssef1313) [inline src/Features/VisualBasic/Portable/ExtractMethod/VisualBasicSelectionValidator.vb:497]: Isn't `Kind` here a method? If so, I think better to change it to `Kind()` or use `IsKind()`.

File: src/Features/VisualBasic/Portable/ExtractMethod/VisualBasicSelectionValidator.vb (hunk 1 above)
Comment (akhera99) [reply src/Features/VisualBasic/Portable/ExtractMethod/VisualBasicSelectionValidator.vb]: oops

File: src/Features/VisualBasic/Portable/ExtractMethod/VisualBasicSelectionValidator.vb (hunk 2 above)
Comment (akhera99) [reply src/Features/VisualBasic/Portable/ExtractMethod/VisualBasicSelectionValidator.vb]: forgot about isKind, thanks!

File: src/Features/VisualBasic/Portable/ExtractMethod/VisualBasicSelectionValidator.vb (hunk 1 above)
Comment (davidwengier) [reply src/Features/VisualBasic/Portable/ExtractMethod/VisualBasicSelectionValidator.vb]: Show me a C# developer, and I'll show you a developer who puts parentheses around their if statements in VB.

File: src/Features/CSharp/Portable/ExtractMethod/CSharpSelectionValidator.cs [hunk 3]
```diff
@@ -205,6 +205,18 @@ private SelectionInfo GetInitialSelectionInfo(SyntaxNode root, SourceText text)
                 };
             }
 
+            var commonRootContainer = commonRoot.FirstAncestorOrSelf<SyntaxNode>(node => node is ClassDeclarationSyntax or RecordDeclarationSyntax or StructDeclarationSyntax);
```
Comment (davidwengier) [inline src/Features/CSharp/Portable/ExtractMethod/CSharpSelectionValidator.cs:208]: Could probably get away with `TypeDeclarationSyntax`, which is the base for all three of these, and `InterfaceDeclarationSyntax`. Its valid to have methods in interfaces too (in C# 8 and above anyway).

**Comment by CyrusNajmabadi** [comment]
Bonjour!

File: src/EditorFeatures/CSharpTest/CodeActions/ExtractMethod/ExtractMethodTests.cs [hunk 4]
```diff
@@ -4289,6 +4289,16 @@ private int GetField()
     {
         return this.field;
     }
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsExtractMethod)]
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/CodeActions/ExtractMethod/ExtractMethodTests.cs:4295]: can you add a workitem to the bug that is fixed?

File: src/Features/CSharp/Portable/ExtractMethod/CSharpSelectionValidator.cs (hunk 3 above)
Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/ExtractMethod/CSharpSelectionValidator.cs:208]: Agreed on this.

Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/ExtractMethod/CSharpSelectionValidator.cs:208]: (would be good to then add tests for the interface case).

File: src/Features/CSharp/Portable/ExtractMethod/CSharpSelectionValidator.cs [hunk 5]
```diff
@@ -205,6 +205,18 @@ private SelectionInfo GetInitialSelectionInfo(SyntaxNode root, SourceText text)
...
+            if (commonRootContainer is null)
+            {
+                return new SelectionInfo
+                {
+                    Status = new OperationStatus(OperationStatusFlag.None, CSharpFeaturesResources.No_valid_selection_to_perform_extraction),
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/ExtractMethod/CSharpSelectionValidator.cs:213]: i think we might want to condier a better string message here.  SPecifically "Selection was not contained inside a type"

also, do we ahve tests for extracing from local functions inside the root namespace?  we don't want to break that (if that works today. ) if it doesn't work prior to your change, tehn no worries.

File: src/Features/VisualBasic/Portable/ExtractMethod/VisualBasicSelectionValidator.vb [hunk 6]
```diff
@@ -494,6 +494,18 @@ result.ReadOutside().Any(Function(s) Equals(s, local)) Then
                        }
             End If
 
+            Dim commonRootContainer = commonRoot.FirstAncestorOrSelf(Of SyntaxNode)(Function(node) node.IsKind(SyntaxKind.ClassBlock) Or node.IsKind(SyntaxKind.StructureBlock) Or node.IsKind(SyntaxKind.ModuleBlock))
+
+            If commonRootContainer Is Nothing Then
+                Return New SelectionInfo With
+                    {
+                        .Status = New OperationStatus(OperationStatusFlag.None, VBFeaturesResources.No_valid_selection_to_perform_extraction),
```
Comment (CyrusNajmabadi) [inline src/Features/VisualBasic/Portable/ExtractMethod/VisualBasicSelectionValidator.vb:502]: this string seems to be duplciated across languages for some reason.  can you move into FeaturesResources and remove from C# and VB.

File: src/Features/CSharp/Portable/ExtractMethod/CSharpSelectionValidator.cs (hunk 5 above)
Comment (ryzngard) [reply src/Features/CSharp/Portable/ExtractMethod/CSharpSelectionValidator.cs:213]: > "Selection was not contained inside a type" 

This may not be a problem for top level statements right?

> "Also, do we ahve tests for extracing from local functions inside the root namespace?"

Is that valid?

Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/ExtractMethod/CSharpSelectionValidator.cs:213]: yup.  top level functions are a thing :)

Comment (akhera99) [reply src/Features/CSharp/Portable/ExtractMethod/CSharpSelectionValidator.cs]: top level functions seem to have been added in C# 9? How can I test if it works?

Comment (akhera99) [reply src/Features/CSharp/Portable/ExtractMethod/CSharpSelectionValidator.cs]: There are no tests for extracting from the root namespace, and I tested it and it has the same error as the bug linked.

**Comment by ryzngard** [comment]
I should take a look at this soon

File: src/Features/CSharp/Portable/ExtractMethod/CSharpSelectionValidator.cs [hunk 7]
```diff
@@ -186,19 +186,45 @@ private SelectionInfo GetInitialSelectionInfo(SyntaxNode root, SourceText text)
...
+            if (commonRootContainer is null && rootGlobalStatement is null)
+            {
+                return new SelectionInfo
+                {
+                    Status = new OperationStatus(OperationStatusFlag.None, FeaturesResources.Selection_not_contained_inside_a_type),
+                    OriginalSpan = adjustedSpan,
+                    FirstTokenInOriginalSpan = firstTokenInSelection,
+                    LastTokenInOriginalSpan = lastTokenInSelection
+                };
+            }*/
```
Comment (ryzngard) [inline src/Features/CSharp/Portable/ExtractMethod/CSharpSelectionValidator.cs:221]: Remove this

File: src/Features/CSharp/Portable/ExtractMethod/Extensions.cs [hunk 8]
```diff
@@ -68,12 +68,21 @@ public static bool UnderValidContext(this SyntaxNode node)
...
+
+            return false;
+
             bool predicate(SyntaxNode n)
             {
-                if (n is BaseMethodDeclarationSyntax ||
-                    n is AccessorDeclarationSyntax ||
-                    n is BlockSyntax ||
-                    n is GlobalStatementSyntax)
+                if (n is BaseMethodDeclarationSyntax or AccessorDeclarationSyntax or BlockSyntax or GlobalStatementSyntax)
```
Comment (ryzngard) [inline src/Features/CSharp/Portable/ExtractMethod/Extensions.cs:85]:
```suggestion
                if (n is BaseMethodDeclarationSyntax or 
                    AccessorDeclarationSyntax or 
                    BlockSyntax or 
                    GlobalStatementSyntax)
```

File: src/EditorFeatures/CSharpTest/ExtractMethod/ExtractMethodTests.cs [hunk 9]
```diff
@@ -9188,13 +9188,13 @@ private static K NewMethod()
         public async Task Script_ArgumentException()
         {
             var code = @"using System;
-public static void GetNonVirtualMethod<TDelegate>( Type type, string name)
+static void GetNonVirtualMethod<TDelegate>( Type type, string name)
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/ExtractMethod/ExtractMethodTests.cs:9191]: please don't update test cases.  instead still record the change in behavior for this case, and add a new test for the differnet case you carea bout.

File: src/Features/CSharp/Portable/ExtractMethod/Extensions.cs [hunk 10]
```diff
@@ -68,12 +68,21 @@ public static bool UnderValidContext(this SyntaxNode node)
...
 
+            if (!node.GetAncestorsOrThis<SyntaxNode>().Any(predicate))
+            {
+                return false;
+            }
+
+            if (node.FromScript() || node.GetAncestor<TypeDeclarationSyntax>() != null)
+            {
+                return true;
+            }
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/ExtractMethod/Extensions.cs:79]: these checks need comments.

File: src/Features/CSharp/Portable/ExtractMethod/Extensions.cs [hunk 11]
```diff
@@ -85,18 +94,48 @@ bool predicate(SyntaxNode n)
...
                 return false;
             }
+        }
 
-            if (!node.GetAncestorsOrThis<SyntaxNode>().Any(predicate))
+        public static bool ContainedInValidType(this SyntaxNode node)
+        {
+            Contract.ThrowIfNull(node);
+
+            if (IsContainedDirectlyInNamespace(node))
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/ExtractMethod/Extensions.cs:103]:
```suggestion
            return !IsContainedDirectlyInNamespace(node);
```

File: src/Features/CSharp/Portable/ExtractMethod/Extensions.cs [hunk 12]
```diff
@@ -85,18 +94,48 @@ bool predicate(SyntaxNode n)
...
+                    if (n is BaseMethodDeclarationSyntax
+                       or AccessorDeclarationSyntax
+                       or BlockSyntax
+                       or CompilationUnitSyntax)
+                    {
+                        hasContainer = true;
+                        immediateParent = true;
+                    }
+
+                    n = n.Parent!;
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/ExtractMethod/Extensions.cs:134]: i do not know what is going on here.  in english what are you trying to express :)

File: src/Features/VisualBasic/Portable/xlf/VBFeaturesResources.zh-Hant.xlf [hunk 13]
```diff
@@ -332,21 +332,6 @@
         <target state="translated">選取範圍無效</target>
         <note />
       </trans-unit>
-      <trans-unit id="Selection_doesn_t_contain_any_valid_token">
```
Comment (CyrusNajmabadi) [inline src/Features/VisualBasic/Portable/xlf/VBFeaturesResources.zh-Hant.xlf:335]: can you make the loc change in a separate PR?

File: src/EditorFeatures/CSharpTest/ExtractMethod/ExtractMethodTests.cs (hunk 9 above)
Comment (akhera99) [reply src/EditorFeatures/CSharpTest/ExtractMethod/ExtractMethodTests.cs:9191]: to be honest, I do not know why I even touched this method

File: src/Features/CSharp/Portable/ExtractMethod/Extensions.cs (hunk 12 above)
Comment (akhera99) [reply src/Features/CSharp/Portable/ExtractMethod/Extensions.cs:137]: if I am a method, or a block, etc. and my immediate parent is a namespace, then bail and do not allow that as a valid selection. Still need to make sure I am handling everything that shouldn't be directly in a namespace

Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/ExtractMethod/Extensions.cs:137]: would it suffice to just ask: did i hit a namespace, but i did not hit a type?  that seems simpler.  or, put another way:

first look if we're contained by a type.  if so, great.
then look to see if we're contained by a namespace.  if so, bad.
otherwise, we must be global.  if so, great.

File: src/Features/VisualBasic/Portable/xlf/VBFeaturesResources.zh-Hant.xlf (hunk 13 above)
Comment (akhera99) [reply src/Features/VisualBasic/Portable/xlf/VBFeaturesResources.zh-Hant.xlf:335]: #55291

File: src/Features/CSharp/Portable/ExtractMethod/Extensions.cs [hunk 14]
```diff
@@ -85,18 +97,24 @@ bool predicate(SyntaxNode n)
 
                 return false;
             }
+        }
 
-            if (!node.GetAncestorsOrThis<SyntaxNode>().Any(predicate))
+        public static bool ContainedInValidType(this SyntaxNode node)
+        {
+            Contract.ThrowIfNull(node);
+            var typeContainer = node.FirstAncestorOrSelf<TypeDeclarationSyntax>();
```
Comment (ryzngard) [inline src/Features/CSharp/Portable/ExtractMethod/Extensions.cs:105]:
```suggestion
           foreach (var ancestor in node.AncestorsAndSelf())
           {
               if (ancestor is TypeDeclarationSyntax) 
                   return true;
               
               if (ancestor is NamespaceDeclarationSyntax)
                   return false;
           }
           
           return true;
```

**Review by ryzngard — APPROVED** [review]

**Comment by ryzngard** [comment]
Current changes look good, pending test passing and conflict resolution
